src/Gaffer/ContextQuery.cpp

```diff
@@ -0,0 +1,342 @@
+
+#include "Gaffer/PlugAlgo.h"
+#include "Gaffer/TypedObjectPlug.h"
+
+#include "IECore/NullObject.h"
+
+#include "boost/bind.hpp"
+
+using namespace Imath;
+using namespace IECore;
+using namespace Gaffer;
+
+namespace
+{
+
+const size_t g_existsPlugIndex = 0;
+const size_t g_valuePlugIndex = 1;
+
+/// \todo: Can the next function be move to somewhere to be shared with `AttributeQuery`?
+
+const Gaffer::ValuePlug *correspondingPlug(
+	const Gaffer::ValuePlug *parent,
+	const Gaffer::ValuePlug *child,
+	const Gaffer::ValuePlug *other
+)
+{
+	boost::container::small_vector< const Gaffer::ValuePlug*, 4 > path;
+
+	const Gaffer::ValuePlug *plug = child;
+
+	while( plug != parent )
+	{
+		path.push_back( plug );
+		plug = plug->parent< Gaffer::ValuePlug >();
+	}
+
+	plug = other;
+
+	while( ! path.empty() )
+	{
+		plug = plug->getChild< Gaffer::ValuePlug >( path.back()->getName() );
+		path.pop_back();
+	}
+
+	return plug;
+}
+
+void addChildPlugsToAffectedOutputs( const Gaffer::Plug* plug, Gaffer::DependencyNode::AffectedPlugsContainer& outputs )
+{
+	if( plug->children().empty() )
+	{
+		outputs.push_back( plug );
+	}
+	else
+	{
+		for( const Gaffer::PlugPtr& child : Gaffer::Plug::OutputRange( *plug ) )
+		{
+			addChildPlugsToAffectedOutputs( child.get(), outputs );
+		}
+	}
+}
+
+/// Returns the child of `parentPlug` that has `descendantPlug` as a descendant or is descendantPlug.
+/// Throws an Exception if the `descendantPlug` is not a descendant of `parentPlug`.
+const Gaffer::ValuePlug *getChildWithDescendant( const Gaffer::Plug *parentPlug, const Gaffer::ValuePlug *descendantPlug )
+{
+	const ValuePlug *p = descendantPlug;
+	while( p )
+	{
+		if( p->parent() == parentPlug )
+		{
+			return p;
+		}
+		p = p->parent<ValuePlug>();
+	}
+
+	throw IECore::Exception( "ContextQuery : Plug not in hierarchy." );
+}
+
+}  // namespace
+
+namespace Gaffer
+{
+
+GAFFER_NODE_DEFINE_TYPE( ContextQuery );
+
+size_t ContextQuery::g_firstPlugIndex = 0;
+
+ContextQuery::ContextQuery( const std::string &name ) : Gaffer::ComputeNode( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new ArrayPlug( "queries" ) );
+	addChild( new ArrayPlug( "out", Plug::Direction::Out ) );
+}
+
+ContextQuery::~ContextQuery()
+{
+}
+
+Gaffer::ArrayPlug *ContextQuery::queriesPlug()
+{
+	return getChild<ArrayPlug>( g_firstPlugIndex + 0 );
+}
+
+const Gaffer::ArrayPlug *ContextQuery::queriesPlug() const
+{
+	return getChild<ArrayPlug>( g_firstPlugIndex + 0 );
+}
+
+Gaffer::ArrayPlug *ContextQuery::outPlug()
+{
+	return getChild<ArrayPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::ArrayPlug *ContextQuery::outPlug() const
+{
+	return getChild<ArrayPlug>( g_firstPlugIndex + 1 );
+}
+
+Gaffer::NameValuePlug *ContextQuery::addQuery(
+	const Gaffer::ValuePlug *plug,
+	const std::string &parameter
+)
+{
+	NameValuePlugPtr childQueryPlug = new NameValuePlug(
+		"",
+		plug->createCounterpart( "query", Gaffer::Plug::Direction::In ),
+		"query1",
```

Comment:
If so, should it be applied to `ShaderQuery` as well? My original reasoning there was naming the user-facing plugs a little more naturally, but `query0` is probably familiar to most people writing expressions and such...

Reply:
My eventual aim is to eliminate names entirely for the children of array plugs, so we only use `[index]` to access them. In the meantime, I think there's value in making sure that `array[N]` and `array["queryN"]` agree on what N is.

Technically I guess applying this to ShaderQuery would be a breaking change (although it shouldn't affect serialisation, because we _do_ use indices there). So if you could make that change for `main` @ericmehl, that'd be great.